src/modules/flow/iio/nodes.c

```diff
@@ -0,0 +1,168 @@
+
+#include <sol-iio.h>
+
+struct gyroscope_data {
+    struct sol_flow_node *node;
+    struct sol_iio_device *device;
+    struct sol_iio_channel *channel_x;
+    struct sol_iio_channel *channel_y;
+    struct sol_iio_channel *channel_z;
+    bool buffer_enabled : 1;
+};
+
+static void
+reader_cb(struct sol_iio_device *device, struct sol_iio_buffer *buffer, void *data)
+{
+    const char errmsg[] = "Could not read channel buffer values";
+    struct gyroscope_data *mdata = data;
+    struct sol_direction_vector out;
+    bool b;
+
+    b = sol_iio_read_buffer_channel_value(mdata->channel_x, buffer, &out.x);
+    if (!b) goto error;
+    b = sol_iio_read_buffer_channel_value(mdata->channel_y, buffer, &out.y);
+    if (!b) goto error;
+    b = sol_iio_read_buffer_channel_value(mdata->channel_z, buffer, &out.z);
+    if (!b) goto error;
+
+    // TODO find appropriate min & max (probably, real min e max outputs)
+    sol_flow_send_direction_vector_packet(mdata->node,
+        SOL_FLOW_NODE_TYPE_IIO_GYROSCOPE__OUT__OUT, &out);
+
+    return;
+
+error:
+    sol_flow_send_error_packet(mdata->node, EIO, errmsg);
+    SOL_WRN(errmsg);
+}
+
+static int
+gyroscope_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct gyroscope_data *mdata = data;
+    const struct sol_flow_node_type_iio_gyroscope_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options, SOL_FLOW_NODE_TYPE_IIO_GYROSCOPE_OPTIONS_API_VERSION,
+        -EINVAL);
+    opts = (const struct sol_flow_node_type_iio_gyroscope_options *)options;
+
+    mdata->buffer_enabled = opts->buffer_size.val > -1;
+
+    mdata->device = sol_iio_open(opts->iio_device.val, opts->buffer_size.val,
+        opts->iio_trigger_name);
+    SOL_NULL_CHECK(mdata->device, -EINVAL);
+
+    mdata->channel_x = sol_iio_add_channel(mdata->device, "in_anglvel_x");
+    SOL_NULL_CHECK_GOTO(mdata->channel_x, error);
+    mdata->channel_y = sol_iio_add_channel(mdata->device, "in_anglvel_y");
+    SOL_NULL_CHECK_GOTO(mdata->channel_y, error);
+    mdata->channel_z = sol_iio_add_channel(mdata->device, "in_anglvel_z");
+    SOL_NULL_CHECK_GOTO(mdata->channel_z, error);
+
+    if (!opts->use_device_default_scale) {
+        sol_iio_set_channel_scale(mdata->channel_x, opts->scale_x.val);
+        sol_iio_set_channel_scale(mdata->channel_y, opts->scale_y.val);
+        sol_iio_set_channel_scale(mdata->channel_z, opts->scale_z.val);
+    }
+    if (!opts->use_device_default_offset) {
+        sol_iio_set_channel_offset(mdata->channel_x, opts->offset_x.val);
+        sol_iio_set_channel_offset(mdata->channel_y, opts->offset_y.val);
+        sol_iio_set_channel_offset(mdata->channel_z, opts->offset_z.val);
+    }
+
+    if (mdata->buffer_enabled) {
+        sol_iio_set_reader_cb(mdata->device, reader_cb, mdata);
+        sol_iio_device_start_buffer(mdata->device);
+    }
+
+    mdata->node = node;
+
+    return 0;
+
+error:
+    SOL_WRN("Could not open IIO device");
```

Comment:
Probably, more warnings were issued before this one, on `sol-iio.c`. Here, probably I should say that 'couldn't create node', and reason is 'could not open iio device %d' - or %s, if not only integers ids will be accepted.